plugins/chat/assets/javascripts/discourse/services/chat-message-visibility-observer.js

```diff
@@ -29,11 +38,27 @@ export default class ChatMessageVisibilityObserver extends Service {
     });
   }
 
+  @bind
+  _mutationObserverCallback(mutationList) {
+    mutationList.forEach((mutation) => {
+      const data = mutation.target.dataset;
+      if (data.id && data.visible && !data.stagedId) {
+        this.chat.updateLastReadMessage();
+      }
+    });
+  }
+
   observe(element) {
-    this.observer.observe(element);
+    this.intersectionObserver.observe(element);
+    this.mutationObserver.observe(element, {
+      attributes: true,
+      attributeOldValue: true,
+      attributeFilter: ["data-staged-id"],
+    });
   }
 
   unobserve(element) {
-    this.observer.unobserve(element);
+    this.intersectionObserver.unobserve(element);
+    this.mutationObserver.disconnect();
```

Comment:
I believe we only want to call `disconnect` when we're sure there are no other elements to observe. I think this function turns off the observer completely.